Make `SetUniqueTogether` idempotent

lieut-data opened the pull request:
In mattermost-server, we effect schema changes across all open SQL connections, e.g.: https://github.com/mattermost/mattermost-server/blob/f9015a37f3f3ffe9dac9d3c3a44c7795b8b8b8b0/store/sqlstore/channel_store.go#L68-L75

This makes sense when there are read replicas, as we may have to apply the schema changes to MySQL instances where they don't automatically propagate from master. Fetching all the SQL stores consists of returning an array containing the master and any replicas: 
https://github.com/mattermost/mattermost-server/blob/71c9dff7662868770f66ab876ad66b354133c2c1/store/sqlstore/supplier.go#L705-L710

Unfortunately, if there aren't any replicas, the `master` connection gets added to the replicas array: 
https://github.com/mattermost/mattermost-server/blob/71c9dff7662868770f66ab876ad66b354133c2c1/store/sqlstore/supplier.go#L242-L244

Effectively resulting in our running through the schema twice against a single master instance. This is fine, since most schema changes are idempotent... except `SetUniqueTogether`. This resulted in duplicate unique indices being created for the `Channels`, `Emoji` and `OAuthAccessData` tables.

I'll be submitting a separate PR to mattermost-server to remove the duplicate indices in a schema migration, and may look at `master` being "incorrectly" part of the `replicas` table, though I suspect it may be easiest just to avoid returning duplicate connections from `GetAllConns` vs. changing the semantics of `replicas`.

lieut-data committed:
fix inability to run gorp tests

lieut-data committed:
make SetUniqueTogether idempotent

It shouldn't be possible to create multiple indexes asserting the same
unique constraints.

lieut-data commented on table.go:
```diff
@@ -87,11 +88,33 @@ func (t *TableMap) SetUniqueTogether(fieldNames ...string) *TableMap {
+	sortedColumns := make([]string, 0, len(fieldNames))
 	for _, name := range fieldNames {
 		columns = append(columns, name)
+		sortedColumns = append(sortedColumns, name)
```
I was inclined just to sort `columns`, but that might result in gorp picking a new name for the index (it seems to use the name of the first key), and recreating a new index on top of an existing one.

Honestly, I think the right medium-term approach is to hand-write the SQL migrations and schema ourselves, including support for +1/-1. It's a bit of overhead when adding columns, but it's then extremely obvious what changes we're making to the database vs. relying on buggy abstraction layers.

jwilander commented on table.go:
```diff
@@ -87,11 +88,33 @@ func (t *TableMap) SetUniqueTogether(fieldNames ...string) *TableMap {
+
+	alreadyExists := false
+	sort.Strings(sortedColumns)
+checkDuplicates:
```
I must say I didn't even know Go had labels.. learn something new everyday

csduarte commented:
A related, but slightly off topic question. Has the results of a duplicate index been tested with performance. Would it cause much impact?

jespino commented on gorp_test.go:
```diff
@@ -831,6 +831,37 @@ func TestSetUniqueTogether(t *testing.T) {
+	dbmap := newDbMap()
+	table := dbmap.AddTable(UniqueColumns{}).SetUniqueTogether("FirstName", "LastName")
+	table.SetUniqueTogether("FirstName", "LastName")
+	table.SetUniqueTogether("LastName", "FirstName")
```
I think it can work for us, but I think in the SQL semantics an Index (unique or not) A,B is not the same than B,A. Actually we can be introducing bugs in other places where that unique index are being used in an specific order.

lieut-data replied:
I agree that these would be different indexes, with different usage semantics. In our case, it seems like we're only using the "unique" part of this given that we then go to the effort of creating other indexes for the columns in question, but agree that this could be confusing. (Another reason to manage the SQL ourselves, IMO!)

We don't need the out-of-order handling, strictly speaking, so I'll remove it to avoid future confusion.

csduarte reviewed (approved):
Change makes sense, test looks good. Can't say I know about usage of gorp to know the cascading effects of this change. -- is gorp still maintained? should this go back upstream?

lieut-data committed:
skip checking sorted unique columns to deduplicate

While it's unlikely one would create multiple unique indexes by specifying
columns in varying orders (better to create named indexes instead), it
could be confusing to automatically deduplicate. Thus, SetUniqueTogether
now only deduplicates using columns in the exact order provided.

lieut-data commented:
@csduarte, re: load testing, no, these changes have not yet been measured for impact, but I'll do that once the changes are merged. Using the force, I suspect memory would benefit primarily, which itself might translate to performance improvements by evicting things from cache less often, but I doubt insertion/update would be impacted all that much.

Internally, we've been discussing moving away from gorp due to lack of activity, but there have been recent commits. I'll look into submitting the patch upstream for good measure regardless.

lieut-data committed:
only ResetSql if adding a unique index

jespino reviewed (approved):
:+1:

lieut-data commented:
I've filed https://github.com/go-gorp/gorp/pull/372 to fix the issue upstream.
